Quotient/threadview.h

```diff
@@ -0,0 +1,49 @@
+// SPDX-FileCopyrightText: 2024 James Graham <james.h.graham@protonmail.com>
+// SPDX-License-Identifier: LGPL-2.1-or-later
+
+#pragma once
+
+#include <QString>
+#include <QJsonObject>
+
+#include "quotient_export.h"
+
+namespace Quotient {
+
+class RoomMessageEvent;
+
+class QUOTIENT_API Thread {
+public:
+    explicit Thread(QString threadRootId, QString latestEventId, int size, bool localUserParticipated);
+
+    QString threadRootId() const;
+    QString latestEventId() const;
+    int size() const;
+    bool localUserParticipated() const;
+
+    bool addEvent(const RoomMessageEvent* event, bool isLatest, bool isLocalUser);
+
+private:
+    const QString _threadRootId;
+    QString _latestEventId;
+    int _size;
+    bool _localUserParticipated;
+};
+
+class QUOTIENT_API ThreadView {
+public:
+    ThreadView() = default;
+
+    void add(std::unique_ptr<Thread> thread);
+
+    bool erase(const QString& threadRootId);
+
+    bool exisits(const QString& threadRootId) const;
+
+    Thread* getThread(const QString& threadRootId) const;
+
+private:
+    std::vector<std::unique_ptr<Thread>> _threads;
```

Comment:
Why `std::unique_ptr` and not `Thread` itself? I don't see polymorphism being involved, and I'm not sure what other reasons might require dynamic allocation of a structure the size of 4 pointers.

Reply:
So currently the Thread object is not copy assignable by design so this makes erasing possible. Maybe however this is just a sign that `std::vector` is not the right container

Reply:
Copy assignability should not be a problem for `std::vector`, it can deal with move-only objects.

Reply:
After quickly playing around with code - so far it looks like you can make `ThreadView` simply a `QHash<QString, Thread>` and implement all the logic you have in `ThreadView` inside `Room::Private::updateThread()`. `Thread` can be a simple `struct` (`Q_GADGET`?). That won't look quite nice when you expose threads outside of `Room`; but if you really need to add the thread root id you can have a separate plain structure for that (possibly but not necessarily derived from `Thread`). In fact, the library already has a similar case with `FileTransferInfo` and `FileTransferPrivateInfo` structures.

Reply:
This is what I ended up with after playing with the code, for what it's worth. This is the meat of `Room::Private::updateThread()`, assuming that `threads` is now `QHash<QString, Thread>` (potentially aliased as `ThreadView` for API convenience), `Thread` is a plain `struct` with public fields, and that pending events are admitted in the thread (which is not happening yet, as `doSendEvent()` doesn't update threads).

I can see that most of this code can be pulled away to `Thread::addEvent()` again, with necessary datapoints from `Room` passed as parameters but I'll leave it at that. My main point is that you don't really need `ThreadView` as a separate class (so far, at least) - once you have found a thread by its root id, the rest is internal to the thread class.
```cpp
    auto& thread = threads[rme->threadRootEventId()];
    thread._localUserParticipated |= rme->senderId() == connection->userId();
    if (++thread._size == 1) { // The thread has just been created
        // For pending events we can get the full correct details when the remote echo comes in.
        thread._latestEventId = rme->id().isEmpty() ? rme->threadRootEventId() : rme->id();
        // When we can't find the root we assume its a historical event that will load later if
        // we can find it we assume a new thread was just created.
        if (!rme->id().isEmpty() && eventsIndex.contains(rme->threadRootEventId()))
            ++thread._size; // Add the thread root event to the size number
    } else {
        const auto threadLatestIndex = eventsIndex.constFind(thread._latestEventId);
        if (threadLatestIndex == eventsIndex.cend()) {
            // Assume this is the latest event. This shouldn't happen but we can work around it.
            thread._latestEventId = rme->id().isEmpty() ? rme->threadRootEventId() : rme->id();
            return;
        }

        const auto eventIndexIt = eventsIndex.constFind(rme->id());
        if (QUO_ALARM_X(
                eventIndexIt == eventsIndex.cend(),
                rme->id()
                    + u"not in the timeline. Update a thread after moving the event to timeline."_s))
            return;

        if (*eventIndexIt > *threadLatestIndex)
            thread._latestEventId = rme->id();
    }
```

Reply:
Yeah, I would totally expect `Thread` getting fatter as we go further. And I think it's worth keeping `ThreadView`, if only as an alias, because I can imagine we might want to change its storage type and/or add functionality to it, too.